Amanda/608 shipment tracking

AmandaRiu opened the pull request:
Fixes #608 by adding read-only support for the Shipment Tracking plugin. If tracking information is available for an order, that information will be printed to the screen, along with a button to track the package. Some highlights:
- Multiple shipment tracking records supported
- If no shipment tracking items, the card will not be displayed
- Loading skeleton not needed since we only show the card after we have data
- Unit tests have been added

<img src="https://user-images.githubusercontent.com/5810477/55442044-a1299380-5572-11e9-96dc-4a6a21d599ca.gif" width="300"/>


Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

nbradbury commented:
@AmandaRiu @kyleaparker I'm wondering if we can reduce the amount of white space here? There's an awful lot of wasted space, which means more scrolling is required in an already long list.

![Screenshot_1554288688](https://user-images.githubusercontent.com/3903757/55473695-437e6100-55dd-11e9-9a45-9a059270050b.png)

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailPresenter.kt:
```diff
@@ -283,6 +315,16 @@ class OrderDetailPresenter @Inject constructor(
+    /**
+     * Fetch the order shipment tracking records from the device db.
+     */
+    fun fetchAndLoadShipmentTrackingsFromDb() {
```
Makes sense. I've updated both the shipment trackings and order notes methods in 1e14984bc34d0557607a743e812a02e96a6cfaad

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailPresenter.kt:
```diff
@@ -225,6 +243,20 @@ class OrderDetailPresenter @Inject constructor(
+                            mapOf(AnalyticsTracker.KEY_ID to order.remoteOrderId))
+
+                    isUsingCachedShipmentTrackings = false
+                    val trackings = orderStore.getShipmentTrackingsForOrder(order)
```
Fixed in fb79c3ebebaa4bb8d8a743a163b437a68947abf3

AmandaRiu commented:
@nbradbury I've updated the design of this card with some changes as discussed earlier in the day:
- Click to copy tracking number to clipboard
- Add shipment date
- Rearrange the layout to consolidate the views
- Since there is no "loading" card, I added an animation when displaying the card so it would be apparent to the user that something changed.

Here's a preview:
<img src="https://user-images.githubusercontent.com/5810477/55517185-3c833d00-5635-11e9-9ab0-589982c4936d.gif" width="300"/>

I fixed the tests as well, though I was sad to find out that Mockito doesn't really care about a generic type invocation specific. For example, we had this line of code:
```
verify(dispatcher, times(1)).dispatch(any<Action<FetchOrderNotesPayload>>())
```
which logically should only count the amount of times an `Action` of type `FetchOrderNotesPayload` is dispatched. However, this is the test error:
```
org.mockito.exceptions.verification.TooManyActualInvocations: 
dispatcher.dispatch(
    <any org.wordpress.android.fluxc.annotations.action.Action>
);
Wanted 1 time:
-> at org.wordpress.android.fluxc.Dispatcher.dispatch(Dispatcher.java:35)
But was 2 times. Undesired invocation:
-> at com.woocommerce.android.ui.orders.OrderDetailPresenter.requestShipmentTrackingsFromApi(OrderDetailPresenter.kt:341)
```
You can see Mockito was looking specifically for just `Action` and didn't care about the rest in this part of the error:
```
dispatcher.dispatch(
    <any org.wordpress.android.fluxc.annotations.action.Action>
);
```
So these were failing because now `loadOrderDetail()` dispatches the request to fetch notes, AND fetch shipment trackings. I've updated the code to look for two generic `Action` dispatches instead in cc723e615657b932d52925fbe17a2e17552e5bb1

Ready for another round!

nbradbury commented:
This looks great! One thing, though: I think it would be better if the tracking info appeared below the customer info instead of above it. The way it is now, the user sees the customer info and then it gets moved down to make room for the tracking info after the request completes. I find that interruptive - I'm looking at the order and then suddenly other info pops in. If the tracking info was lower, the user wouldn't see things moving around like this.

AmandaRiu commented:
@nbradbury moved the card below customer info in edc43469da3bb061737223ede1b40321384c7be4. I wonder though if we should move it below the payment information since that seems to be more of the linear expectation of the order: Status -> products -> customer -> payment -> tracking. Thoughts @kyleaparker ?

nbradbury commented:
> I wonder though if we should move it below the payment information since that seems to be more of the linear expectation of the order: Status -> products -> customer -> payment -> tracking.

That makes sense to me.

kyleaparker commented:
Yep makes sense 👍

AmandaRiu commented:
Awesome! updated and ready for final round @nbradbury 👍

nbradbury commented:
Looks great, very nice work! :shipit:
